### PR TITLE
[profile] MRG: auto-profile improvement

### PR DESCRIPTION
**Pull request by YuanmingLeee**
Closes #

#### Before submitting this PR, please make sure you have:

- [x] run `python -m pytest tests/` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited.

Progress:
- [x] Profile result save to MongoDB
- [x] Auto profile at the background (submit profile job to the controller when registering a model, rather than profile right at the registration time)
- [x] Status check

Test:
- [x] TFS Client
- [x] ONNX Client
- [x] TorchScript Client
- [x] TRT Client

### CODE REVIEW

**Comment by HuaizhengZhang** [comment]
How about the progress of this pr?

**Comment by YuanmingLeee** [comment]
> How about the progress of this pr?

- [x] Profile result save to MongoDB
- [ ] Auto profile at the background (submit profile job to the controller when registering a model, rather than profile right at the registration time)

Test:
- [x] TFS Client
- [ ] ONNX Client
- [ ] TorchScript Client
- [ ] TRT Client

Current Issue:
1. Packed services need time to startup, need a status check for `profiler` / `controller`
2. `controller` as a profiling job scheduler, to support profile in the background.

**Comment by huangyz0918** [comment]
@YuanmingLeee can you finish up here first?

image

For testing, I think you have successfully gone through TFS, other things can be done in a less priority.

**Comment by HuaizhengZhang** [comment]
Cheers!

Hurry up!

File: modelci/hub/deployer/serving.py [hunk 1]
```diff
@@ -1,6 +1,4 @@
 import argparse
```
Comment (HuaizhengZhang) [inline modelci/hub/deployer/serving.py:1]: image
why delete this? It seems good
```
from .serving import serve as dispatcher

__all__ = ['dispatcher']
```

File: modelci/hub/client/README.md [hunk 2]
```diff
@@ -68,11 +68,11 @@ test_img_bytes = ...
...
 tfs_client = CVTFSClient(test_img_bytes, batch_num=100, batch_size=32, asynchronous=False)
 # get the model info from model manager.
-mode_info = retrieve_model(architecture_name='ResNet50', framework=Framework.PYTORCH, engine=Engine.TORCHSCRIPT)
+mode_info = retrieve_model(architecture_name='ResNet50', framework=Framework.TENSORFLOW, engine=Engine.TFS)
 # init a profiler.
-profiler = Profiler(model_info=mode_info, server_name='tfs', inspector=tfs_client)
+profiler = Profiler(model_bo=mode_info, server_name='tfs', inspector=tfs_client)
 # start profiling.
-profiler.diagnose('device name')
+profiler.diagnose('cuda:0')
```
[the remark this answers is not in the document]
Comment (huangyz0918) [reply modelci/hub/client/README.md:75]: I think @YuanmingLeee he just passing the cuda id so that in later we can use the GPUtil to get the brand name. This looks good to me since in the future there may be serving in CPU devices.

File: README.md [hunk 3]
```diff
@@ -125,7 +124,7 @@ model_info = ...
 torch_client = CVTorchClient(test_data_item, batch_num, batch_size, asynchronous=False)
 
 # init the profiler
-profiler = Profiler(model_info=model_info, server_name='name of your server', inspector=torch_client)
+profiler = Profiler(model_bo=model_info, server_name='name of your server', inspector=torch_client)
```
[the remark this answers is not in the document]
Comment (huangyz0918) [reply README.md:127]: I think we can keep name `model_info` and comment here to indicate it is a data object here. what do ur think? @YuanmingLeee

File: modelci/hub/client/tfs_client.py [hunk 4]
```diff
@@ -1,37 +1,55 @@
...
+        self.signature_name = signature_name
         super().__init__(repeat_data=repeat_data, batch_num=batch_num, batch_size=batch_size, asynchronous=asynchronous)
+
         self.request = None
         self.stub = None
 
-    def data_preprocess(self):
-        pass
+    def data_preprocess(self, x):
+        return cv2.resize(x, tuple(self.inputs[0].shape[1:3])).astype(np.float32)
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply modelci/hub/client/tfs_client.py:41]: from TFS official client script,
```shell script
img = tf.keras.preprocessing.image.load_img(file, target_size=[224, 224])
x = tf.keras.preprocessing.image.img_to_array(img)
```
The first line: `file_path` -> `<PIL Image>` with resize to 224 x 224
The second line: `<PIL Image>` -> `numpy.ndarray, dtype=float32`

As we already passed in a `ndarray` with any size, we use `cv2::resize` to compress the image into given size.
The size info is stored in `inputs`. `inputs[0].shape[1:3]` just fetch the width and heigh info of the first input arg of the model.
`.astype` is used for converting the `ndarray` type to `float32` as required.

File: modelci/hub/client/tfs_client.py [hunk 5]
```diff
@@ -1,37 +1,55 @@
...
-        tf.app.flags.DEFINE_string('image', './data/cat.jpg', 'path to image in JPEG format')
+    def __init__(
+            self,
+            repeat_data,
+            model_name: str,
+            inputs: List[IOShape],
+            batch_num=1,
+            batch_size=1,
+            asynchronous=None,
+            signature_name: str = 'serving_default',
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply modelci/hub/client/tfs_client.py:30]: Given by TFS client official script:
https://github.com/tensorflow/serving/blob/102261586fac6bcc28d03071767c83ef36c9bf96/tensorflow_serving/example/resnet_client_grpc.py#L55

File: modelci/hub/deployer/serving.py (hunk 1 above)
Comment (YuanmingLeee) [reply modelci/hub/deployer/serving.py:1]: this will raise the circular reference issue.
```
serving deps: manager
manger deps: profiler
profiler deps: serving
```
Cannot put in `__init__.py`. Could you suppose a better structure?

File: modelci/hub/client/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply modelci/hub/client/README.md:75]: Yep, GPUtil will get the device name and UUID via device string: e.g. `cuda:0`, `cpu`.

File: README.md [hunk 6]
```diff
@@ -99,8 +99,7 @@ from modelci.types.bo import IOShape
 ONNXConverter.from_torch_module(
     '<path to torch model>', 
     '<path to export onnx model>', 
-    input_shape=[IOShape([-1, 3, 224, 224], float)], 
-    batch_size=16
+    inputs=[IOShape([-1, 3, 224, 224], float)],
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply README.md:102]: batch size is indicated at `input_shape`. The first dim is batch size.

File: docs/tutorial/profile.md [hunk 7]
```diff
@@ -16,18 +16,24 @@ We provide many build-in gRPC clients:
...
-```python 
+```python
 from modelci.hub.client.torch_client import CVTorchClient
 from modelci.hub.profiler import Profiler
 
+test_data_item = ...
+batch_num = 100
+batch_size = 32
+
+model_bo = ...
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply docs/tutorial/profile.md:27]: Suggest a name? As this arg is using the type `model_bo`. @huangyz0918

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply README.md:127]: Ok, let’s stick to `model_info`

File: README.md [hunk 8]
```diff
@@ -140,7 +139,7 @@ MLModelCI provides a dispatcher to deploy a model as a cloud service. The dispat
 We search for a converted model and then dispatch it to a device with a specific batch size.
 
 ```python
-from modelci.hub.deployer import serve
+from modelci.hub.deployer.serving import serve
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply README.md:142]: dispatcher

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (HuaizhengZhang) [reply README.md:102]: better add comments here

File: modelci/hub/deployer/serving.py (hunk 1 above)
Comment (HuaizhengZhang) [reply modelci/hub/deployer/serving.py:1]: Add a todo, we will fix this up later.

File: modelci/hub/profiler.py [hunk 9]
```diff
@@ -47,16 +51,28 @@ def __init__(self, model_bo: ModelBO, server_name: str, inspector: BaseModelInsp
...
-        """Start diagnosing and profiling model."""
-
-        try:  # to check the container has started successfully or not.
-            self.docker_client.containers.get(self.server_name)
-        except Exception:
-            print(
-                '\n'
-                'ModelCI Error: starting the serving engine failed, please start the Docker container manually. \n'
-            )
+    def diagnose(self, batch_size: int = None, device='cuda', timeout=10) -> DynamicProfileResultBO:
```
Comment (HuaizhengZhang) [inline modelci/hub/profiler.py:54]: device = "cuda"

seems there is no clear way for CPU?

what if profiling models on CPU

File: modelci/hub/client/tfs_client.py [hunk 10]
```diff
@@ -4,49 +4,65 @@
...
         self.request = predict_pb2.PredictRequest()
         self.request.model_spec.name = self.model_name
         self.request.model_spec.signature_name = self.signature_name
 
+    def check_model_status(self) -> bool:
+        api_url = f'http://{self.SERVER_HOST}:{TFS_HTTP_PORT}/v1/models/{self.model_name}/versions/{self.version}'
+        try:
+            response = requests.get(api_url)
+            state = response.json()['model_version_status'][0]['state']
+            return state == 'AVAILABLE'
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply modelci/hub/client/tfs_client.py:58]: this status is obtained from TF serving. See this API: https://www.tensorflow.org/tfx/serving/api_rest#model_status_api

[the remark this answers is not in the document]
Comment (HuaizhengZhang) [reply modelci/hub/client/tfs_client.py:58]: Good. Let us the official definition.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply README.md:102]: See docs in ONNXConverter https://github.com/YuanmingLeee/ML-Model-CI/blob/0cb19ee10d05666f07fc4d3197e5bd0ada61f9b5/modelci/hub/converter.py#L60

File: modelci/hub/profiler.py (hunk 9 above)
Comment (YuanmingLeee) [reply modelci/hub/profiler.py:54]: Current profiler fails in profiling on CPU @huangyz0918:
https://github.com/YuanmingLeee/ML-Model-CI/blob/0cb19ee10d05666f07fc4d3197e5bd0ada61f9b5/modelci/metrics/benchmark/metric.py#L132
This line fails, as there is no key named `'accelerators'` on CPU. Please help to fix this after this merge.

Comment (huangyz0918) [reply modelci/hub/profiler.py:54]: I know this issue, it comes from cadvisor. Will fix, not hurry.

Comment (huangyz0918) [reply modelci/hub/profiler.py:54]: Remember do not increase the scope of a current working pull request. Try to finalize the required changes ASAP.

Comment (HuaizhengZhang) [reply modelci/hub/profiler.py:54]: add a todo better
to remind of us and developers

Comment (huangyz0918) [reply modelci/hub/profiler.py:54]: Yuanming has filed an issue for that @HuaizhengZhang

File: modelci/hub/deployer/serving.py (hunk 1 above)
Comment (YuanmingLeee) [reply modelci/hub/deployer/serving.py:1]: Done

File: modelci/hub/deployer/serving.py [hunk 11]
```diff
@@ -54,18 +52,21 @@ def serve(
         environment['CUDA_VISIBLE_DEVICES'] = device_num
 
     if engine == Engine.TFS:
+        docker_tag = '2.1.0-gpu' if cuda else '2.1.0'
```
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply modelci/hub/deployer/serving.py:56]: Done

File: modelci/hub/profiler.py [hunk 12]
```diff
@@ -36,37 +48,72 @@ def __init__(self, model_info, server_name: str, inspector: BaseModelInspector =
                 raise TypeError("The inspector should be an instance of class BaseModelInspector!")
 
         self.server_name = server_name
-        self.model_info = model_info
+        self.model_bo = model_info
```
Comment (HuaizhengZhang) [inline modelci/hub/profiler.py:51]: 在hub的代码中，变量名尽量不出现bo，po之类的吧
感觉跳来跳去不清晰（虽然我已经习惯了）

Comment (HuaizhengZhang) [reply modelci/hub/profiler.py:51]: bo这些在pesistence里面出现就行

File: modelci/metrics/benchmark/metric.py [hunk 13]
```diff
@@ -39,17 +54,17 @@ def __init__(self, repeat_data, batch_num=1, batch_size=1, asynchronous=False, p
 
         self.batch_num = batch_num
         self.batch_size = batch_size
+        self.model_info = model_info
```
Comment (HuaizhengZhang) [inline modelci/metrics/benchmark/metric.py:57]: 对啊，这就很舒服

File: scripts/install.conda_env.sh [hunk 14]
```diff
@@ -1,15 +1,12 @@
...
 
 # Activate conda
-source "${HOME}"/anaconda3/etc/profile.d/conda.sh
+source "${CONDA_PREFIX}"/etc/profile.d/conda.sh
 conda activate modelci
 
 # Install PyTorch and TensorFlow package
-conda install pytorch=1.5.0 torchvision cudatoolkit="${CUDA_VERSION}" -y -c pytorch
-conda install tensorflow-gpu=2.1.0 -y
+conda install pytorch=1.5.0 torchvision cpuonly -y -c pytorch
```
Comment (HuaizhengZhang) [inline scripts/install.conda_env.sh:10]: ？
why cpu only？

一直用的pytorch 1.5？那这里严格先固定，先不要动了

File: modelci/hub/client/tfs_client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HuaizhengZhang) [reply modelci/hub/client/tfs_client.py]: 我总感觉.shape[1:3]这个切片在这里怪怪的，加个todo吧，之后我再看看

File: scripts/install.conda_env.sh (hunk 14 above)
Comment (YuanmingLeee) [reply scripts/install.conda_env.sh:10]: modelci don't need to run any model directly on host machine, we only need some util function, such as `torchvision`, model export. All running part is inside docker. Therefore, we need not CUDA support

File: modelci/hub/profiler.py (hunk 12 above)
Comment (YuanmingLeee) [reply modelci/hub/profiler.py:51]: ok, all outside use `business object`. Data object is for persistence, and View object is for jsonify

File: scripts/install.conda_env.sh (hunk 14 above)
Comment (HuaizhengZhang) [reply scripts/install.conda_env.sh:10]: I know. But cpuonly only supports CPU while gpu supports both.

From any view, directly installing GPU version is better than  CPU version.

（不要钻牛角尖，这个改进没有必要啊）

Comment (YuanmingLeee) [reply scripts/install.conda_env.sh:10]: TF don't support cuda 10.2 (our install script not working on 10.2 machine). We can fix all CUDA version issue at one hit.

Comment (HuaizhengZhang) [reply scripts/install.conda_env.sh:10]: 我不是很理解你想表达的点，我们这一行和10.2有什么关系呢？

Comment (huangyz0918) [reply scripts/install.conda_env.sh:10]: 李老师的意思应该是，ModelCI 其实不需要 GPU 支持的 TensorFlow 和 PyTorch，因为不需要训练，只是用一下其中的一些 util 和 type 类。TensorFlow 现阶段不支持 cuda 10.2 所以干脆直接改成 CPU only，免得在一些机器上安装失败

Comment (HuaizhengZhang) [reply scripts/install.conda_env.sh:10]: 这就没必要了，我们cuda最好固定成10.1，方便做论文，如果是这个原因，那也ok把，之后改

File: modelci/hub/client/tfs_client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (YuanmingLeee) [reply modelci/hub/client/tfs_client.py]: changed

File: modelci/hub/client/tfs_client.py [hunk 15]
```diff
@@ -45,9 +45,9 @@ def __init__(
         self.signature_name = signature_name
 
     def data_preprocess(self, x):
-        shape = self.model_info.inputs[0].shape
-        dtype = model_data_type_to_np(self.model_info.inputs[0].dtype)
-        return cv2.resize(x, tuple(shape[1:3])).astype(dtype)
+        input_ = self.model_info.inputs[0]
+        dtype = model_data_type_to_np(input_.dtype)
+        return cv2.resize(x, (input_.height, input_.width)).astype(dtype)
```
Comment (HuaizhengZhang) [inline modelci/hub/client/tfs_client.py:50]: put some comments or references?

Comment (HuaizhengZhang) [reply modelci/hub/client/tfs_client.py:50]: Then MRG

**Comment by huangyz0918** [comment]
please fix the code styles

File: modelci/hub/client/tfs_client.py (hunk 15 above)
Comment (YuanmingLeee) [reply modelci/hub/client/tfs_client.py:52]: Added.
